roofit/roostats/src/AsymptoticCalculator.cxx

```diff
@@ -1421,17 +1416,15 @@ RooAbsData * AsymptoticCalculator::MakeAsimovData(const ModelConfig & model, con
       }
       // unfold the nuisance pdf if it is a prod pdf
       RooArgList pdfList;
-      RooProdPdf *prod = dynamic_cast<RooProdPdf *>(nuispdf.get());
+      RooProdPdf *prod = static_cast<RooProdPdf *>(nuispdf.get());
       if (prod ) {
          pdfList.add(prod->pdfList());
       }
       else
          // nothing to unfold - just use the pdf
          pdfList.add(*nuispdf.get());
 
-      RooLinkedListIter iter(pdfList.iterator());
-      for (RooAbsArg *a = (RooAbsArg *) iter.Next(); a != 0; a = (RooAbsArg *) iter.Next()) {
-         RooAbsPdf *cterm = dynamic_cast<RooAbsPdf *>(a);
+      for (auto *cterm : static_range_cast<RooAbsPdf *>(pdfList)) {
```

Comment:
By using `static_cast` here, you make the `assert` below ineffective. I suggest you do the `dynamic_cast` here inside the assert`, so it's not active in non-debug builds. So please change the line below to:
```C++
assert(dynamic_cast<RooAbsPdf*>(cterm) && "AsimovUtils: a factor of the nuisance pdf is not a Pdf!");
```